### PR TITLE
feature(http): adds a service for signing and validating URLs

### LINKED ISSUE #9884: Signed URLs
I am thinking about a way we could sign and validate URLs. Some use cases would include uservalidationbyemail, certain actions, confirmation links in notifications etc.

I am thinking we could add something like `elgg_sign_url()`, which would parse query parameters, remove `__elgg_ts` and `__elgg_token`, cast values to string, sort by key, and then add HMAC signature as `_elgg_mac` to the URL. `elgg_validate_url_signature()` would then validate query parameters of the given URL.

Thoughts?

### PR DESCRIPTION
**Pull request by hypeJunction**
URLs can now be signed with a HMAC key and validated at a later stage using
elgg_http_get_signed_url() and elgg_http_validate_signed_url().

Fixes #9884

### CODE REVIEW

**Comment by hypeJunction** [comment]
@ewinslow This would make #8647 quite easy to implement.

**Comment by hypeJunction** [comment]
@mrclay When you get a chance, I would like to hear your feedback on this. Any security implications that I missing?

File: engine/classes/Elgg/Di/ServiceProvider.php [hunk 1]
```diff
@@ -60,6 +60,7 @@
  * @property-read \Elgg\Router                             $router
  * @property-read \Elgg\Application\ServeFileHandler       $serveFileHandler
  * @property-read \ElggSession                             $session
+ * @property-read \Elgg\Security\SignedUrl                 $signedUrl
```
Comment (mrclay) [inline engine/classes/Elgg/Di/ServiceProvider.php]: UrlSigner / $urlSigner ?

**Comment by mrclay** [comment]
LGTM, though it might make sense to migrate something to use this as a test. Like email validation or forgot password.

File: engine/classes/Elgg/Security/SignedUrl.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	 *                        Falsey values indicate non-expiring URL
+	 * @return string
+	 */
+	public function sign($url, $expires = false) {
+		$url = elgg_normalize_url($url);
+		
+		$parts = parse_url($url);
+
+		if (isset($parts['query'])) {
+			$query = elgg_parse_str($parts['query']);
```
Comment (mrclay) [inline engine/classes/Elgg/Security/SignedUrl.php]: You might throw here if `__elgg_mac` is already set.

**Comment by hypeJunction** [comment]
@mrclay, it didn't make sense getting rid of the code in the password reset because it's used in a form, so just added another layer there. Uservalidationbyemail now uses the new API.

File: engine/classes/Elgg/PasswordService.php [hunk 3]
```diff
@@ -89,7 +89,8 @@ function sendNewPasswordRequest($user_guid) {
 
 		// generate link
 		$link = _elgg_services()->config->getSiteUrl() . "changepassword?u=$user_guid&c=$code";
```
Comment (mrclay) [inline engine/classes/Elgg/PasswordService.php]: We don't need the $code anymore.

Comment (hypeJunction) [inline engine/classes/Elgg/PasswordService.php]: I didn't want to hack my way through the password service. The code is injected in the password reset form and then validated. There would be BC implications removing it entirely.

File: engine/lib/elgglib.php [hunk 4]
```diff
@@ -1293,6 +1293,41 @@ function elgg_http_url_is_identical($url1, $url2, $ignore_params = array('offset
...
+function elgg_http_validate_signed_url($url) {
+	return _elgg_services()->urlSigner->isValid($url);
+}
+
+/**
+ * Validates if the HMAC signature of the current request is valid
+ * Issues 403 response if signature is inalid
+ * @return void
+ */
+function elgg_http_validate_request_signature() {
```
Comment (mrclay) [inline engine/lib/elgglib.php]: I assume a "validate" function will return a bool. Maybe `elgg_assert_valid_signed_request` or `elgg_signed_request_gatekeeper`?

Comment (hypeJunction) [inline engine/lib/elgglib.php]: I like `elgg_signed_url_gatekeeper()`

**Comment by mrclay** [comment]
We might want to add caveats that signed URLs are **not bound to the session** and shouldn't be used for actions in general. E.g. We might be tempted to replace action URLs with these to avoid the hassle of validating the parameters, but this would allow a leaked URL to be used by others, and possibly replayed; the permissions system is a bit of a fail-safe here, but it isn't everything.

It *would* be safe, however, to sign full action URLs and use the signed URL gatekeeper function at the top of the action.

**Comment by hypeJunction** [comment]
@mrclay See if this covers it.

**Comment by mrclay** [comment]
LGTM

**Comment by hypeJunction** [comment]
It's been two weeks, merging.

**Comment by Wouter0100** [comment]
Muh.. This change isn't really BC. You've just deleted the `c` input parameter instead of deprecated? Shouldn't these things get removed in a BC-breaking release? (e.g. 3.x, 4.x).

**Comment by hypeJunction** [comment]
What does it break?

**Comment by Wouter0100** [comment]
The old verification method with the c parameter.

**Comment by hypeJunction** [comment]
Why do you need the c parameter? URL is signed and verified.

**Comment by hypeJunction** [comment]
It's internal plugin logic, and you shouldn't rely on system generated URLs to stay the same.

**Comment by mrclay** [comment]
I'm sorry this hit you. It can be hard to identify the edges of the API, but we now lean towards progress over my paranoid conservatism.

**Comment by Wouter0100** [comment]
Hmm, yeah, I can understand it may be an edge case. In the case system generated URLs will change, we could still have it BC compatible, by providing a general method for creating system generated URLs (like possible with routes). For example, in [this](https://github.com/Centillien/validation_reminder) plugin we need to send an e-mail (different contents) with the validation URL (https://github.com/Centillien/validation_reminder/commit/4f460ab2132352472bca17d56bcd274d40d5f2bd), but due the change it wasn't working properly.

**Comment by hypeJunction** [comment]
We just need to add a wrapper `uservalidationbyemail_get_validation_url()`

**Comment by Wouter0100** [comment]
@hypeJunction yep. Something like that.

**Comment by hypeJunction** [comment]
Open an issue please.

**Pull request merged**